File: ch04/22/IsPopOrder.cpp
// 给两个序列，第一个序列表示栈的压入顺序，请判断第二个序列是否为该栈的弹出顺序
class Solution {
public:
    bool IsPopOrder(vector<int> pushV,vector<int> popV) {
        bool result = false;
        int push_len = pushV.size();
        int pop_len = popV.size();
        if (push_len != pop_len)
           return result;

        stack<int> pushS;
        int p1 = 0; 
        int p2 = 0; // 如果这个指针遍历了popV，就说明是弹出序列
        while (true) {
            if ( p1 < push_len) {
                if (pushV[p1] != popV[p2] ) {
                    pushS.push(pushV[p1]);
                    p1 ++;
                }
                else {
                    p2 ++;
                    p1 ++;
                }
            }
            else {
                if (!pushS.empty()) {
                    if (pushS.top() == popV[p2]) {
                        pushS.pop();
                        p2 ++;
                    }else {
                        result = false;
                        break;
                    }
                }else {
                    result = true;
                    break;
                }
            }
        }
        return result;
    }
};

File: ch04/22/IsPopOrder.py
# -*- coding:utf-8 -*-
class Solution:
    def IsPopOrder(self, pushV, popV):
        # write code here
        result = False
        push_len = len(pushV)
        pop_len = len(popV)
        if push_len != pop_len:
            return result

        p1 = 0
        p2 = 0
        stack = []
        while True:
            if p1 < push_len:
                if pushV[p1] != popV[p2]:
                    stack.append(pushV[p1])
                    p1 += 1
                else:
                    p1 += 1
                    p2 += 1
            else:
                if len(stack) != 0:    
                    if stack[len(stack) - 1] == popV[p2]:
                        stack.pop()
                        p2 += 1
                    else:
                        result =  False
                        break
                else:
                    result = True
                    break

        return result  